pkg/controller/operators/catalog/scratch.go

```diff
@@ -0,0 +1,383 @@
+package catalog
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/client"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/clientset/versioned"
+	exv "github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/informers/externalversions"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/registry/reconciler"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorclient"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/operatorlister"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/queueinformer"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/metrics"
+)
+
+// What goes into a view?
+// Set of indexers that generate keys for primary resources.
+// View function that transforms inbound runtime objects into the stored view.
+
+// ViewFunc returns a view and associated view key for an object.
+// type ViewFunc func(obj interface{}) (view interface{}, key string, handles bool, err error)
+
+type View interface {
+	Key(value interface{}) (key string, err error)
+	Value(obj interface{}) (value interface{}, err error)
+}
+
+// Views are a set of ViewFuncs keyed by their resource type
+type Views map[reflect.Type]View
+
+// Viewer is a storage interface that supports building alternate views of stored data.
+type Viewer interface {
+	cache.Indexer
+
+	// View gets the value for a view type and given object.
+	View(viewType reflect.Type, obj interface{}) (value interface{}, err error)
+
+	// AddViews adds more views to this store. If you call this after you already have data
+	// in the store, the results are undefined.
+	AddViews(views Views) error
+}
+
+type viewCache struct {
+	cache.Indexer
+	lock       sync.RWMutex
+	views      Views
+	defaultKey cache.KeyFunc
+}
+
+func (c *viewCache) View(viewType reflect.Type, obj interface{}) (value interface{}, err error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	view, ok := c.views[viewType]
+	if !ok {
+		return nil, fmt.Errorf("view %s not found", viewType.Name())
+	}
+
+	value, err = view.Value(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	key, err := view.Key(value)
```

Comment:
I was thinking that we want to get the value of what's stored in the cache rather than a fresh view.

Reply:
My last comment was wrong, sorry.

What I meant was: 

Why not have an interface for getting a key directly from an object? 

i.e. 

```go
type View interface {
	Key(obj interface{}) (key string, err error)
	Value(obj interface{}) (value interface{}, err error)
}
```
vs
```go
type View interface {
	Key(value interface{}) (key string, err error)
	Value(obj interface{}) (value interface{}, err error)
}
```

so that here, you would just write `view.Key(obj)` and not need to get the value at all.

Reply:
Ah, got it - makes sense. We probably want to add another function for getting the key from the "view" itself (to satisfy some [internal requirements](https://github.com/kubernetes/client-go/blob/65184652c88974f455258f2e0fbc6ac2c67e1786/tools/cache/store.go#L124) of the underlying cache implementation).